pyomo/gdp/plugins/fix_disjuncts.py

```diff
@@ -0,0 +1,109 @@
+from six import itervalues
+
+__author__ = "Qi Chen <https://github.com/qtothec>"
+
+logger = logging.getLogger('pyomo.core')
+
+
+class GDP_Disjunct_Fixer(Transformation):
+    """Fix disjuncts to Blocks."""
+
+    def __init__(self, *args, **kwargs):
+        self.integer_tolerance = kwargs.pop('int_tol', 1E-6)
+        super(GDP_Disjunct_Fixer, self).__init__(*args, **kwargs)
+        self._transformedDisjuncts = ComponentSet()
+
+    alias('gdp.fix_disjuncts',
+          doc=textwrap.fill(textwrap.dedent(__doc__.strip())))
+
+    def _apply_to(self, instance):
+        t = instance
+        if not t.active:
+            return
+
+        if (type(t) not in (_DisjunctData, _BlockData, _DisjunctionData) and
+                t.type() not in (Disjunct, Block, Disjunction)):
+            raise GDP_Error(
+                "Target %s was not a Block, Disjunct, or Disjunction. "
+                "It was of type %s and can't be transformed."
+                % (t.name, type(t)))
+
+        if t.is_indexed():
+            for obj in itervalues(t):
+                self._transformObject(obj)
+        else:
+            self._transformObject(t)
+
+    def _transformObject(self, obj):
+        if type(obj) is _DisjunctionData:
+            self._transformDisjunctionData(obj)
+        self._transformContainer(obj)
+
+    def _transformDisjunctionData(self, disjunction):
+        logical_sum = sum(value(disj.indicator_var)
+                          for disj in disjunction.disjuncts)
+        if disjunction.xor and not logical_sum == 1:
+            raise GDP_Error(
+                "Disjunction %s violated. "
+                "Expected 1 disjunct to be active, but %s were active."
+                % (disjunction.name, logical_sum))
+        elif not logical_sum >= 1:
+            raise GDP_Error(
+                "Disjunction %s violated. "
+                "Expected at least 1 disjunct to be active, "
+                "but none were active.")
+        else:
+            disjunction.deactivate()
+
+        for disj in disjunction.disjuncts - self._transformedDisjuncts:
+            self._transformDisjunctData(disj)
+        self._transformedDisjuncts.update(disjunction.disjuncts)
+
+    def _transformDisjunctData(self, obj):
+        if fabs(value(obj.indicator_var) - 1) <= self.integer_tolerance:
+            # Disjunct is active. Convert to Block.
+            obj.parent_block().reclassify_component_type(obj, Block)
+            obj.indicator_var.fix(1)
+            self._transformContainer(obj)
+        elif fabs(value(obj.indicator_var)) <= self.integer_tolerance:
+            obj.parent_block().reclassify_component_type(obj, Block)
+            obj.indicator_var.fix(0)
+            # Deactivate all constituent constraints and disjunctions
+            for constr in obj.component_objects(
+                    ctype=(Constraint, Disjunction),
+                    active=True, descend_into=True):
+                constr.deactivate()
```

Comment:
Why not just deactivate the entire block?

Reply:
I'm paranoid about the solvers not detecting variables on deactivated blocks.

Reply:
Ack.  You are absolutely correct.  There is a flipside, though: Disjuncts with ActiveComponents other than Constraint that aren't deactivated when you turn the disjunct off.  We really need to file a ticket and force that fix for Pyomo 5.5.  Can you at least document here that this is a HACK to get around the deactivated blocks stupidity in the writers?

Reply:
Sure. It also depends on the outcome of the discussion in https://github.com/Pyomo/pyomo/issues/309.